JSON Schema draft 2020 compatibility addition

M-casado opened the pull request:
## Issue / Ticket Reference
https://github.com/elixir-europe/biovalidator/issues/81

## Overall Changes
- Modified how ``Ajv`` instances are created, dropping the bespoke ``Ajv2019`` and unused ``Ajv2020``. 
- Removed the hardcoded block stopping draft 2020 from working.
- Added skipping snippets for the ``$async`` not to be injected in loops for official JSON Schema drafts when resolving references.
- Updated AJV requirement to above 8.12.0
- Added ``examples/both/`` with both ``data`` and ``schema`` together in a JSON, to test the deployed server through requests. Each draft from 06 to 2020 has a ``valid`` and ``invalid`` file.

## Future To-Dos
- [x] Add proper tests trhough npm test for each draft that Biovalidator is compatible with

## Testing
````
$ npm test

> biovalidator@2.2.2 test
> jest

...

Test Suites: 13 passed, 13 total
Tests:       36 passed, 36 total
Snapshots:   0 total
Time:        23.113 s
Ran all test suites.
````

And a more manual testing to check that drafts are accepted and validation works as intended:
````
$ for file in $(ls ./*-valid*); do curl --data @$file -H "Content-Type: application/json" -X POST "http://localhost:3020/validate" ; echo $file ; done
[]./test-06SchemaDraft-valid.json
[]./test-07SchemaDraft-valid.json
[]./test-2019SchemaDraft-valid.json
[]./test-2020SchemaDraft-valid.json

$ for file in $(ls ./*-invalid*); do curl --data @$file -H "Content-Type: application/json" -X POST "http://localhost:3020/validate" ; echo $file ; done
[{"dataPath":"/foo","errors":["must be equal to constant"]}]./test-06SchemaDraft-invalid.json
[{"dataPath":"/bar","errors":["must have required property 'bar'"]},{"dataPath":"","errors":["must match \"then\" schema"]}]./test-07SchemaDraft-invalid.json
[{"dataPath":"/foo","errors":["must be string"]}]./test-2019SchemaDraft-invalid.json
[{"dataPath":"/foo","errors":["must be string"]}]./test-2020SchemaDraft-invalid.json
````

M-casado committed:
Remove forced block of 2020-12

M-casado committed:
Upgrade Ajv dependency to 8.x for 2020-12 support

M-casado committed:
Use standard Ajv import instead of Ajv2019

M-casado committed:
Avoid fetching and injecting  in draft schemas

M-casado committed:
Add full examples of drafts 06, 07, 2019 and 2020

M-casado committed:
Avoid undefined schema

M-casado committed:
Swap $id with $schema for identifying drafts

theisuru reviewed (dismissed):
Looks good and thanks for refactoring the the code around ajv version handling. I will do little bit more testing with custom keywords and merge. 

One improvement that could be done is adding test cases for the new files in `examples/both`

M-casado committed:
Add drafts Jest tests

M-casado commented:
Added a fourteenth test for the drafts. Now picked by Jest.
````
$ npm test

> biovalidator@2.2.2 test
> jest

...

Test Suites: 14 passed, 14 total
Tests:       44 passed, 44 total
Snapshots:   0 total
Time:        22.049 s, estimated 24 s
Ran all test suites.
````

theisuru commented:
![image](https://github.com/user-attachments/assets/7f394cac-8237-45d3-923b-0dd01eb29021)

I get this error when I ran it locally and use the browser to validate using the `Load Sample Data` button. Do you get the same error?

M-casado commented:
Hi @theisuru, I got the exact same error when I checked it through the UI. After a bit of testing here and there I noticed that the issue was only appearing where **a custom keyword was used**. Then, after checking some of the schemas I was using, I noticed the key difference: this example one **does not contain** ``"$async": true,`` at root.

The moment you add that missing bit at root, it works as expected:
![image](https://github.com/user-attachments/assets/9c5777ea-f73d-46b4-b305-366e195b3e7d)


If I remember correctly, the logic of the script was that "if there is ``"$async": true,`` at root, then the code adds ``"$async": true,`` to every nested property of the schemas, to avoid users adding it themselves". Nevertheless, if this used to work, and now doesn't, it may be that with the new logic of "skip if it's an official draft", now if ``"$async": true,`` is missing at root, then it's not added automatically, and it fails.

From this I take that:

- **We should improve the logs of the tool**. I played around with the log when I was doing changes for this PR, but I dropped the changes as it was unrelated. I think we should add much clearer error logs, especially when these things happen regarding ``Ajv`` expecting something to be asynchronous when it's not, or the other way around.
- We should define whether ``"$async": true,`` **has** to be **added at root by the user**, or not, when using a custom keyword.
- We should **add a test for this in the suite of tests**, given that every test was passing but only when you checked that example, we noticed.

M-casado commented:
To be honest I'm glad that we had this issue, because I just learnt (my bad for not reading the README lately again) about the ``queryFields`` argument, which is **pure-gold** for what I wanted to do with metadata validation.

theisuru commented:
Hi @M-casado sorry for the late reply. 
Yes, I agree. Definitely need improvements in logs and tests areas.  
About the `$async`: At the moment we are adding `$async` to all schemas and `definitions` whether they have it at the top level or not.

M-casado commented:
> About the $async: At the moment we are adding $async to all schemas and definitions whether they have it at the top level or not.

I believe that then I may have changed the logic very slightly:
- If it was a root schema, whose ``$id`` corresponded to the canonical JSON Schema drafts, there were two options:
   - If there's a ``$async`` at root, it adds ``$async`` to everything under it
   - If there's no ``$async`` at root, then since it's a canonical schema it skips it (to avoid endless addition of ``$async``)

For this PR to go further, would you say changing this logic again would be necessary? Or just to add some tests for this to be picked up?

M-casado committed:
Refactor injection of '' when not found at root

M-casado committed:
Add asyncCustomKeywords.test.js to test '' behaviour at root

M-casado committed:
Refactor core JS adding new logging statements

M-casado commented:
Hi @theisuru, I added some of the changes you were requesting, and refactored slightly the code, adding more logging statements.

The only thing left is for me to check again the CHANGELOG. I need to update it, since I added yet another ``npm`` test module, and a couple of example JSONs. After that, 📣 **I think it's on your plate to accept it and merge it** 📣

## Testing
### NPM test modules
````
$ npm test

> biovalidator@2.2.2 test
> jest

...

Test Suites: 15 passed, 15 total
Tests:       48 passed, 48 total
Snapshots:   0 total
Time:        23.771 s
Ran all test suites.
````

### UI loaded example
![image](https://github.com/user-attachments/assets/4f9c9817-c2d9-4996-bcc0-57294d83038e)

M-casado commented:
Strangely enough... I just noticed that there's no CHANGELOG in the repo. Then... I guess there's nothing for me to add anymore 💪

M-casado committed:
Improve logging and support ''

M-casado committed:
Improve debug logging

M-casado committed:
Simplify back local schema compilation

M-casado committed:
Remove redundant debug log

M-casado committed:
Improve logger's messages format

M-casado committed:
Remove verbose statement

M-casado committed:
Merge pull request #1 from M-casado/improve-logger

Addition of logger improvements

M-casado commented:
@theisuru - I took the matter into my own hands to also improve the logging and reporting of errors and debug statements. Thus, I created a new branch ``improve-logger`` and added the changes to this branch in a **[new PR](https://github.com/M-casado/biovalidator/pull/1)** for convenience to see the changes with respect to this branch.

As per the previous tests, I checked that all tests passed:
````
$ npm test

> biovalidator@2.2.2 test
> jest

...

Test Suites: 15 passed, 15 total
Tests:       48 passed, 48 total
Snapshots:   0 total
Time:        25.526 s, estimated 29 s
Ran all test suites.
````

Likewise, I tested that the loaded example works in the UI too:
![image](https://github.com/user-attachments/assets/44762207-309c-48f1-ab70-994fb447da19)

With a log example as follows:
````
$ node src/biovalidator
2025-04-23T14:17:27.637Z [info] Custom keywords successfully added. Number of custom keywords: 5
2025-04-23T14:17:27.689Z [info] ---------------------------------------------
2025-04-23T14:17:27.690Z [info] ------------ ELIXIR biovalidator ------------
2025-04-23T14:17:27.691Z [info] ---------------------------------------------
2025-04-23T14:17:27.691Z [info] Started server on port 3020 with base URL /
2025-04-23T14:17:27.691Z [info] Server available at http://localhost:3020/
2025-04-23T14:17:27.692Z [info] PID file is available at ../biovalidator/server.pid
2025-04-23T14:17:27.692Z [info] Writing logs to: ../biovalidator/logs/
2025-04-23T14:18:05.172Z [debug] BioValidator.validate() called for initial schema '$id': '(no '$id')'
2025-04-23T14:18:05.173Z [debug] Auto-injected "$async": true at root for schema '$id': '[no '$id' at root]'
2025-04-23T14:18:05.173Z [debug] Compiling schema '$id': (no '$id'). This will trigger loading of external references.
2025-04-23T14:18:05.173Z [warn] Compiling schema with empty schema '$id'. Schema will not be cached in validator cache.
2025-04-23T14:18:05.179Z [info] Successfully obtained compiled function for schema '$id': '[no $id in schema]'.
2025-04-23T14:18:05.855Z [debug] Returning resolved term from OLS: [http://purl.obolibrary.org/obo/MONDO_0000001,http://purl.obolibrary.org/obo/PATO_0000461]
2025-04-23T14:18:06.197Z [debug] Returning resolved term from OLS: [http://purl.obolibrary.org/obo/MONDO_0000001,http://purl.obolibrary.org/obo/PATO_0000461]
2025-04-23T14:18:06.197Z [info] Validation finished successfully for schema '$id': '[no $id in schema]'.
2025-04-23T14:18:06.199Z [info] New validation request: Processed successfully in 1027ms.
````

As you can see, the improved logging system is, as AJV, heavily dependent on the ``$id`` of the schemas, to report how the compilation of them goes when Biovalidator traverses references. Without ``$id``, it's not as informative. With ``$id``, it helped me debug the issue that started all of this at Beacon: https://github.com/ga4gh-beacon/beacon-v2/issues/206

M-casado commented:
Given the contributors' tracing to this repo, might as well bring @snathanvj into the conversation :)

Feel free to suggest changes, or approve the PR.

theisuru reviewed (approved):
Thanks for improvements, this looks good.
